### ISSUE TITLE
HW intrinsics: Do better with generic method cases

### ISSUE THREAD

**Issue by voinokin** [dotnet/coreclr#19851]
Is there any chance getting generic versions of `Sse2.And/Or/Xor/AndNot()` in API like it was done for `Avx2.BroadcastScalarToVector128<T>()` and `Avx2.BroadcastScalarToVector256<T>()` ? (Not sure about FP variants.) These operations are word size agnostic, as well as `Sse2.ShiftLeft/RightLogical128BitLane()`, `Ssse3.AlignRight()`, `Sse41.TextXXX()`, and maybe some others.

**Comment by fiigii** [comment]
Do you want something like this?

https://github.com/dotnet/coreclr/blob/f9c2e89f31f3816c921abb7c84e1951ce37d3127/tests/src/JIT/HardwareIntrinsics/X86/General/VectorHelpers.cs#L14-L60

**Comment by voinokin** [comment]
Yes, I mean something like this. My use case is - I have algorithm which
does not depend much on data word width, and I'd like to keep it readable
(+/- constants which do depend).

As an example, simple part of it conditionally swaps two vectors and looks
like:

V128<TKey> aXORb = Xor(a, b);
V128<TKey> swapMask = CompareGreaterThan(a, b); // more complex condition
used with type casting and broadcasts
a = Xor(a, And(aXORb, cmp));
b = Xor(b, And(aXORb, cmp));

There are already means to do some simple "vertical" operations like this
in System.Numerics, but then there's no way to exchange data with HW
intrinsics APIs.

**Comment by fiigii** [comment]
>There are already means to do some simple "vertical" operations like this
in System.Numerics, but then there's no way to exchange data with HW
intrinsics APIs.

Hardware intrinsic has different design goal from System.Numerics.Vectors, which hardware intrinsic is a serial of lower level APIs. So, you need to implement your own higher-level/generic APIs based on your application environment/design (e.g., hardware bottom-line, software fallback, etc.). That is why we provide generic `Vector128/256<T>` types (instead of C++'s non-generic vectors) and `StaticCast<T, U>`.

**Comment by voinokin** [comment]
That is understandable, and helper methods similar to your example can be
written for personal ) use. But I was wondering whether .Net Core will
provide some out of the box.

**Comment by tannergooding** [comment]
@voinokin. This is on my list of things to raise next Tuesday (at the next API Design Review Meeting).

However, due to the layering of the x86 APIs (there are very few places where support for all 10 primitives were added at the same time) and due to last weeks decision for the ARM APIs (which was to just explode everything, rather than exposing a single generic method). I don't imagine this is going to end up with us exposing more generic wrappers in box.

I do think exposing a general purpose generic wrapper, as a 3rd part library, is probably a fine idea.

I would also think that exposing more APIs directly on `System.Numerics.Vector<T>` (after following the API Review Process) is something that is reasonable (a number of the APIs you mentioned are already exposed there).
